CS3219-AY2425S1/cs3219-ay2425s1-project-g17
# Add CRUD for user-service, Docker files for user-service and project

## To test the CRUD

1. cd into `user-service` directory
2. Run `docker build . -t cs3219g17/user-service --no-cache`
3. Run `docker run --name usr-svc -p 4001:4001 -d cs3219g17/user-service`
4. You should see the docker container running when you run `docker ps`
5. You can now use Postman to send HTTP requests to port 4001.

## Endpoints:

### Create user: http://localhost:4001/users [`POST`]

- Body 
  - Required: `username` (string), `email` (string), `password` (string)

    ```JSON
    {
      "username": "SampleUserName",
      "email": "sample@gmail.com",
      "password": "SecurePassword"
    }
    ```

### Get all users: http://localhost:4001/users [`GET`]

- Header 
    - Required: `Authorization: Bearer <JWT_ACCESS_TOKEN>`
    - Auth Rules: Only admin users can retrieve all users' data.

### Get a user: http://localhost:4001/users/{userId} [`GET`]

- Parameters
    - Required: `userId` path parameter
    - Example: `http://localhost:4001/users/60c72b2f9b1d4c3a2e5f8b4c`
- Headers
    - Required: `Authorization: Bearer <JWT_ACCESS_TOKEN>`
    - Auth Rules: Admin users can retrieve any user's data, non-admin users can only retrieve their own data.

### Update a user: http://localhost:4001/users/{userId} [`PATCH`]

- Parameters
    - Required: `userId` path parameter
- Body
    - At least one of the following fields is required: `username` (string), `email` (string), `password` (string)

    ```JSON
    {
      "username": "SampleUserName",
      "email": "sample@gmail.com",
      "password": "SecurePassword"
    }
    ```

- Headers
    - Required: `Authorization: Bearer <JWT_ACCESS_TOKEN>`
    - Auth Rules: Admin users can update any user's data, non-admin users can only update their own data.

### Update user privilege: http://localhost:4001/users/{userId} [`PATCH`]

- Parameters
    - Required: `userId` path parameter
- Body
    - Required: `isAdmin` (boolean)
    
    ```JSON
    {
      "isAdmin": true
    }
    ```

- Headers
    - Required: `Authorization: Bearer <JWT_ACCESS_TOKEN>`
    - Auth Rules: Only admin users can update any user's privilege.

### Delete a user: http://localhost:4001/users/{userId} [`DELETE`]

- Parameters
    - Required: `userId` path parameter
- Headers
    - Required: `Authorization: Bearer <JWT_ACCESS_TOKEN>`
    - Auth Rules: Admin users can delete any user's data, non-admin users can only delete their own data.

### Login: http://localhost:4001/auth/login [`POST`]

- Body
    - Required: `email` (string), `password` (string)

    ```JSON
    {
      "email": "sample@gmail.com",
      "password": "SecurePassword"
    }
    ```

### Verify token: http://localhost:4001/auth/verify-token [`GET`]

- Headers
    - Required: `Authorization: Bearer <JWT_ACCESS_TOKEN>`
